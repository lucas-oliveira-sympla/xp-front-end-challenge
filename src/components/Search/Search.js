import React, { Component } from 'react';
import { func } from 'prop-types';

import { FormStyled, SearchStyled, InputStyled } from './Search.styles';

const propTypes = {
  handleSearch: func.isRequired,
};

class Search extends Component {
  constructor() {
    super();

    this.changedHandler = this.changedHandler.bind(this);
  }

  changedHandler(event) {
    const { handleSearch } = this.props;
    const term = event.target.value;
    handleSearch(term);
  }

  render() {
    return (
      <FormStyled action="">
        <SearchStyled htmlFor="search">
          Busque por artistas, álbuns ou músicas
          <InputStyled
            type="text"
            id="search"
            placeholder="Comece a escrever..."
            onChange={this.changedHandler}
          />
        </SearchStyled>
      </FormStyled>
    );
  }
}

Search.propTypes = propTypes;

export default Search;
