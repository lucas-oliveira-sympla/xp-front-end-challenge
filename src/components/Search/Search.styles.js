import styled from 'styled-components';

import colors from '../../theme/colors';

export const FormStyled = styled.form`
  padding: 0 5px 16px;
  border-bottom: solid 1px ${colors.primary[700]};
  margin-bottom: 20px;
`;

export const SearchStyled = styled.label`
  color: ${colors.primary[900]};
  display: block;
  font-family: 'Roboto', sans-serif;
  font-size: 16px;
`;

export const InputStyled = styled.input`
  background-color: transparent;
  border: none;
  color: ${colors.primary[300]};
  display: block;
  font-size: 48px;
  font-weight: bold;
  height: 50px;
  margin-top: 10px;
  width: 100%;
`;
