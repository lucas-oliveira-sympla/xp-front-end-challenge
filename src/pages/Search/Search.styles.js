import styled from 'styled-components';
import breakpoint from 'styled-components-breakpoint';

const MainWrapper = styled.section`
  max-width: 1130px;
  padding: 10px;

  ${breakpoint('desktop')`
    margin: 75px auto 0;
  `}

  > .list {
    display: flex;
    flex-flow: row wrap;
    justify-content: flex-start;

    ${breakpoint('tablet')`
      margin: 0 -30px;
    `}

    > a {
      min-height: 260px;
      margin-top: 40px;
      transition: transform 0.2s linear;
      flex: 0 0 50%;
      max-width: 50%;
      padding: 0 10px;

      ${breakpoint('tablet')`
        flex: 0 0 25%;
        max-width: 25%;
        padding: 0 30px;
      `}

      ${breakpoint('desktop')`
        flex: 0 0 20%;
        max-width: 20%;
      `}

      &:hover {
        transition: transform 0.1s linear;
        transform: scale(1.1);
      }
    }
  }
`;

export default MainWrapper;
