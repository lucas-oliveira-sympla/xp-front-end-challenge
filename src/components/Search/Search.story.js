import React from 'react';
import { storiesOf } from '@storybook/react';
import styled from 'styled-components';
import colors from '../../theme/colors';

import Search from './Search';

const BackgroundStyled = styled.div`
  background-color: ${colors.primary[900]};
  padding: 5%;
`;

const stories = storiesOf('Search', module);
stories.add('default', () => (
  <BackgroundStyled>
    <Search />
  </BackgroundStyled>
));
