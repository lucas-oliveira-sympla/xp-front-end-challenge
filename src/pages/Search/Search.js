import React, { Component } from 'react';
import {
  oneOfType, shape, func, string,
} from 'prop-types';
import { connect } from 'react-redux';
import axios from 'axios';
import slugify from 'slugify';

import * as actionTypes from '../../store/actions';
import Search from '../../components/Search';
import Heading from '../../components/Heading';
import Card from '../../components/Card';

import MainWrapper from './Search.styles';

const propTypes = {
  token: oneOfType([
    shape({}),
    shape({
      accessToken: string.isRequired,
    }),
  ]),
  searchTerm: string,
  unsetToken: func,
  setSearchTerm: func,
  requestSearch: func,
  receiveSearch: func,
};

const defaultProps = {
  token: null,
  searchTerm: '',
  unsetToken: () => {},
  setSearchTerm: () => {},
  requestSearch: () => {},
  receiveSearch: () => {},
};

class SearchPage extends Component {
  constructor() {
    super();

    this.searchContent = this.searchContent.bind(this);
    this.renderAlbums = this.renderAlbums.bind(this);
  }

  searchContent(term) {
    const {
      token: { accessToken },
      unsetToken,
      setSearchTerm,
      requestSearch,
      receiveSearch,
    } = this.props;

    if (!term) {
      receiveSearch(null);
      return;
    }

    requestSearch();

    const queryParams = [];
    queryParams.push(`q=${encodeURIComponent(term)}`);
    queryParams.push('type=album,artist,track');

    const options = {
      headers: {
        Authorization: `Bearer ${accessToken}`,
      },
    };

    axios
      .get(`https://api.spotify.com/v1/search?${queryParams.join('&')}`, options)
      .then((response) => {
        const { albums } = response.data;

        setSearchTerm(term);
        receiveSearch(albums);
      })
      .catch((error) => {
        console.error('ERROR', error);
        unsetToken();
      });
  }

  renderAlbums() {
    const { albums } = this.props;
    if (!albums) return null;

    const cards = albums.items.map((item, index) => {
      const cardProps = {
        key: slugify(`${index}${item.name}-${item.artists[0].name}`),
        imgUrl: item.images[1].url,
        album: item.name,
        artist: item.artists[0].name,
      };

      return <Card {...cardProps} />;
    });

    return cards;
  }

  renderHeading() {
    const { searchTerm } = this.props;
    if (searchTerm) {
      return (
        <Heading>
          Resultado da busca por
          {` "${searchTerm}"`}
        </Heading>
      );
    }
    return <Heading>√Ålbuns buscados recentemente</Heading>;
  }

  render() {
    return (
      <>
        {/* <Logo src={SpotifyLogo} alt="Spotify Logo" /> */}
        <MainWrapper>
          <section className="search">
            <Search handleSearch={this.searchContent} />
          </section>
          {this.renderHeading()}
          <section className="list">{this.renderAlbums()}</section>
        </MainWrapper>
      </>
    );
  }
}

SearchPage.propTypes = propTypes;
SearchPage.defaultProps = defaultProps;

const mapStateToProps = (state) => ({
  isAuthenticated: state.isAuthenticated,
  token: state.token,
  searchTerm: state.searchTerm,
  albums: state.albums,
});

const mapDispatchToProps = (dispatch) => ({
  unsetToken: () => dispatch({ type: actionTypes.UNSET_TOKEN }),
  setSearchTerm: (searchTerm) => dispatch({
    type: actionTypes.SET_SEARCH_TERM,
    searchTerm,
  }),
  requestSearch: () => dispatch({ type: actionTypes.REQUEST_SEARCH }),
  receiveSearch: (albums) => dispatch({
    type: actionTypes.RECEIVE_SEARCH,
    albums,
  }),
});

export default connect(mapStateToProps, mapDispatchToProps)(SearchPage);
