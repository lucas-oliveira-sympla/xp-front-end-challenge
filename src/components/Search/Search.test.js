import React from 'react';
import { mount } from 'enzyme';

import Search from './Search';

describe('<Search />', () => {
  let wrapper;

  describe('with default props', () => {
    beforeAll(() => {
      wrapper = mount(<Search handleSearch={() => {}} />);
    });

    it('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
