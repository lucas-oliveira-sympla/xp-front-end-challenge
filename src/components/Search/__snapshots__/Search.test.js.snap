// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Search /> with default props should match snapshot 1`] = `
<Search
  handleSearch={[Function]}
>
  <Searchstyles__FormStyled
    action=""
  >
    <StyledComponent
      action=""
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "Searchstyles__FormStyled-sc-1d5ecam-0",
            "isStatic": false,
            "lastClassName": "cUcevQ",
            "rules": Array [
              "border-bottom:solid 1px ",
              "#FAFAFA",
              ";padding:0 0 5px;margin-bottom:10px;",
              [Function],
            ],
          },
          "displayName": "Searchstyles__FormStyled",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "Searchstyles__FormStyled-sc-1d5ecam-0",
          "target": "form",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
    >
      <form
        action=""
        className="Searchstyles__FormStyled-sc-1d5ecam-0 cUcevQ"
      >
        <Searchstyles__SearchStyled
          htmlFor="search"
        >
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "Searchstyles__SearchStyled-sc-1d5ecam-1",
                  "isStatic": false,
                  "lastClassName": "cnabGP",
                  "rules": Array [
                    "color:",
                    "#FFF",
                    ";display:block;font-family:'Roboto',sans-serif;font-size:16px;",
                  ],
                },
                "displayName": "Searchstyles__SearchStyled",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "Searchstyles__SearchStyled-sc-1d5ecam-1",
                "target": "label",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
            htmlFor="search"
          >
            <label
              className="Searchstyles__SearchStyled-sc-1d5ecam-1 cnabGP"
              htmlFor="search"
            >
              Busque por artistas, álbuns ou músicas
              <Searchstyles__InputStyled
                autoComplete="off"
                id="search"
                onKeyUp={[Function]}
                placeholder="Comece a escrever..."
                type="search"
              >
                <StyledComponent
                  autoComplete="off"
                  forwardedComponent={
                    Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "attrs": Array [],
                      "componentStyle": ComponentStyle {
                        "componentId": "Searchstyles__InputStyled-sc-1d5ecam-2",
                        "isStatic": false,
                        "lastClassName": "nQjor",
                        "rules": Array [
                          "background-color:transparent;border:none;color:",
                          "#999",
                          ";display:block;font-weight:bold;height:50px;margin-top:10px;width:100%;font-size:24px;",
                          [Function],
                        ],
                      },
                      "displayName": "Searchstyles__InputStyled",
                      "foldedComponentIds": Array [],
                      "render": [Function],
                      "styledComponentId": "Searchstyles__InputStyled-sc-1d5ecam-2",
                      "target": "input",
                      "toString": [Function],
                      "warnTooManyClasses": [Function],
                      "withComponent": [Function],
                    }
                  }
                  forwardedRef={null}
                  id="search"
                  onKeyUp={[Function]}
                  placeholder="Comece a escrever..."
                  type="search"
                >
                  <input
                    autoComplete="off"
                    className="Searchstyles__InputStyled-sc-1d5ecam-2 nQjor"
                    id="search"
                    onKeyUp={[Function]}
                    placeholder="Comece a escrever..."
                    type="search"
                  />
                </StyledComponent>
              </Searchstyles__InputStyled>
            </label>
          </StyledComponent>
        </Searchstyles__SearchStyled>
      </form>
    </StyledComponent>
  </Searchstyles__FormStyled>
</Search>
`;
